File: windows/Hangly.Windows.Tests/RopeSimulationTests.cs
using System;
using System.Linq;
using System.Windows;
using Hangly.Windows.Physics;
using Xunit;

namespace Hangly.Windows.Tests;

public class RopeSimulationTests
{
    private readonly Vector2D _anchor = new(260, 15);
    private const double Frame120 = 1.0 / 120.0;

    private RopeSimulation MakeRope()
    {
        var rope = new RopeSimulation(RopeConfiguration.Default, _anchor);
        rope.Start();
        return rope;
    }

    private static void Run(RopeSimulation rope, double seconds)
    {
        int steps = (int)(seconds / Frame120);
        for (int i = 0; i < steps; i++)
        {
            rope.Step(Frame120);
        }
    }

    private static double TotalSpeed(RopeSimulation rope)
    {
        return rope.Points.Sum(p => p.Displacement.Magnitude);
    }

    // MARK: - Structure

    [Fact]
    public void SegmentCountMatchesSpecification()
    {
        var rope = MakeRope();
        Assert.Equal(20, rope.Configuration.SegmentCount);
        Assert.Equal(21, rope.Points.Count);
    }

    [Fact]
    public void MassDistributionIsCorrect()
    {
        var rope = MakeRope();

        Assert.True(rope.Points[0].IsPinned);
        Assert.Equal(0.0, rope.Points[0].InverseMass);
        // A heavier charm has a smaller inverse mass than a plain node.
        Assert.True(rope.Points[20].InverseMass < rope.Points[10].InverseMass);
        Assert.Equal(1.0, rope.Points[10].InverseMass);
    }

    [Fact]
    public void AnchorStaysPinned()
    {
        var rope = MakeRope();
        rope.BeginDrag(rope.Points[20].Position);
        rope.UpdateDrag(new Vector2D(4000, -4000), new Vector2D(9000, -9000));
        Run(rope, 2.0);

        Assert.Equal(_anchor, rope.Points[0].Position);
    }

    // MARK: - Inextensibility

    [Fact]
    public void DoesNotStretchAtRest()
    {
        var rope = MakeRope();
        Run(rope, 5.0);

        Assert.True(rope.MeasuredMaximumStretch <= rope.Configuration.MaxStretchRatio + 1e-9);
    }

    [Fact]
    public void DoesNotStretchUnderHardFlick()
    {
        var rope = MakeRope();
        Run(rope, 1.0);
        rope.BeginDrag(rope.Points[20].Position);

        var position = rope.Points[20].Position;
        double worstStretch = 0.0;

        for (int tick = 0; tick < 1200; tick++)
        {
            double direction = ((tick / 18) % 2 == 0) ? 1.0 : -1.0;
            position += new Vector2D(direction * 3000.0 * Frame120, Math.Sin(tick * 0.05) * 12.0);
            rope.UpdateDrag(position, new Vector2D(direction * 3000.0, 0));
            rope.Step(Frame120);
            worstStretch = Math.Max(worstStretch, rope.MeasuredMaximumStretch);
        }

        Assert.True(worstStretch <= rope.Configuration.MaxStretchRatio + 1e-9);
    }

    [Fact]
    public void DoesNotStretchWhileSwinging()
    {
        var rope = MakeRope();
        Run(rope, 1.0);

        double worstStretch = 0.0;
        for (int i = 0; i < 2400; i++)
        {
            rope.Step(Frame120);
            worstStretch = Math.Max(worstStretch, rope.MeasuredMaximumStretch);
        }

        Assert.True(worstStretch < 1.01);
    }

    [Fact]
    public void RecoversFromUnreachableInput()
    {
        var rope = MakeRope();
        Run(rope, 1.0);
        rope.BeginDrag(rope.Points[20].Position);

        double worstStretch = 0.0;
        for (int tick = 0; tick < 600; tick++)
        {
            double angle = tick * 0.35;
            var target = _anchor + (new Vector2D(Math.Cos(angle), Math.Sin(angle)) * 900.0);
            rope.UpdateDrag(target, new Vector2D(6000.0, 6000.0));
            rope.Step(Frame120);
            worstStretch = Math.Max(worstStretch, rope.MeasuredMaximumStretch);
        }
        Assert.True(worstStretch < 1.05);

        rope.EndDrag();
        Run(rope, 1.0);
        Assert.True(rope.MeasuredMaximumStretch <= rope.Configuration.MaxStretchRatio + 1e-9);
    }

    // MARK: - Gravity and Damping

    [Fact]
    public void GravityPullsTheRopeDown()
    {
        var rope = MakeRope();
        double startX = rope.Points[20].Position.X;
        Run(rope, 20.0);
        var charm = rope.Points[20].Position;

        Assert.True(Math.Abs(charm.X - _anchor.X) < Math.Abs(startX - _anchor.X) * 0.25);
        Assert.True(charm.Y > _anchor.Y + (rope.Configuration.TotalLength * 0.9));
    }

    [Fact]
    public void DampingSettlesTheRope()
    {
        var rope = MakeRope();
        Run(rope, 3.0);
        double movingSpeed = TotalSpeed(rope);

        Run(rope, 25.0);
        double settledSpeed = TotalSpeed(rope);

        Assert.True(settledSpeed < movingSpeed * 0.1);
    }

    // MARK: - Frame-Rate Independence

    [Fact]
    public void FixedTimeStepMakesRefreshRateIrrelevant()
    {
        var fast = new RopeSimulation(RopeConfiguration.Default, _anchor);
        var slow = new RopeSimulation(RopeConfiguration.Default, _anchor);
        fast.Start();
        slow.Start();

        for (int i = 0; i < 120; i++)
        {
            fast.Step(1.0 / 120.0);
            fast.Step(1.0 / 120.0);
            slow.Step(1.0 / 60.0);
        }

        for (int index = 0; index < fast.Points.Count; index++)
        {
            Assert.True(Math.Abs(fast.Points[index].Position.X - slow.Points[index].Position.X) < 1e-9);
            Assert.True(Math.Abs(fast.Points[index].Position.Y - slow.Points[index].Position.Y) < 1e-9);
        }
    }

    [Fact]
    public void RemainsStableOverALongRun()
    {
        var rope = MakeRope();
        Run(rope, 120.0);

        double reach = rope.Configuration.TotalLength * 3.0;
        foreach (var point in rope.Points)
        {
            Assert.True(point.Position.IsFinite);
            Assert.True(point.Position.DistanceTo(_anchor) < reach);
        }
    }

    [Fact]
    public void ClampsOversizedFrames()
    {
        var rope = MakeRope();
        rope.Step(10.0);

        int ceiling = (int)(rope.Configuration.MaxFrameDuration / rope.Configuration.FixedTimeStep) + 1;
        Assert.True(rope.LastStepCount <= ceiling);
    }

    // MARK: - Interaction

    [Fact]
    public void GrabIsLimitedToTheCharm()
    {
        var rope = MakeRope();
        Run(rope, 5.0);

        Assert.True(rope.CanGrab(rope.Points[20].Position));
        Assert.False(rope.CanGrab(_anchor));
        Assert.False(rope.CanGrab(rope.Points[10].Position));
        Assert.False(rope.BeginDrag(_anchor));
        Assert.False(rope.IsDragging);
    }

    [Fact]
    public void DraggingMovesTheCharmToTheCursor()
    {
        var rope = MakeRope();
        Run(rope, 5.0);
        rope.BeginDrag(rope.Points[20].Position);

        var target = new Vector2D(_anchor.X + 120.0, _anchor.Y + 140.0);
        for (int i = 0; i < 10; i++)
        {
            rope.UpdateDrag(target, Vector2D.Zero);
            rope.Step(Frame120);
        }

        Assert.True(rope.Points[20].Position.DistanceTo(target) < 1e-6);
    }

    [Fact]
    public void ReleasePreservesMomentum()
    {
        var rope = MakeRope();
        Run(rope, 5.0);
        rope.BeginDrag(rope.Points[20].Position);

        double radius = rope.Configuration.TotalLength * 0.9;
        double speed = 900.0;
        double angularStep = (speed / radius) * Frame120;
        double angle = Math.PI / 2.0;

        for (int i = 0; i < 16; i++)
        {
            angle -= angularStep;
            var target = _anchor + (new Vector2D(Math.Cos(angle), Math.Sin(angle)) * radius);
            var tangent = new Vector2D(Math.Sin(angle), -Math.Cos(angle)) * speed;
            rope.UpdateDrag(target, tangent);
            rope.Step(Frame120);
        }

        var held = rope.Points[20].Displacement / Frame120;
        rope.EndDrag();
        rope.Step(Frame120);
        var released = rope.Points[20].Displacement / Frame120;

        Assert.True(released.Magnitude > held.Magnitude * 0.5);
        Assert.True((released.X * held.X) + (released.Y * held.Y) > 0);
        Assert.True(released.X > 0);
    }

    [Fact]
    public void ReleaseWithoutMotionDoesNotLaunchTheCharm()
    {
        var rope = MakeRope();
        Run(rope, 5.0);
        rope.BeginDrag(rope.Points[20].Position);

        var held = _anchor + (new Vector2D(0.6, 0.8) * (rope.Configuration.TotalLength * 0.9));
        for (int i = 0; i < 60; i++)
        {
            rope.UpdateDrag(held, Vector2D.Zero);
            rope.Step(Frame120);
        }
        rope.EndDrag();
        rope.Step(Frame120);

        double speed = rope.Points[20].Displacement.Magnitude / Frame120;
        Assert.True(speed < 200.0);
    }

    // MARK: - Idling

    [Fact]
    public void SleepsWhenSettled()
    {
        var rope = MakeRope();
        Assert.False(rope.IsSleeping);

        Run(rope, 40.0);

        Assert.True(rope.IsSleeping);
        Assert.Equal(0, rope.LastStepCount);
    }

    [Fact]
    public void SleepingRopeDoesNotDrift()
    {
        var rope = MakeRope();
        Run(rope, 40.0);
        var resting = rope.Points.Select(p => p.Position).ToList();

        Run(rope, 20.0);

        for (int i = 0; i < rope.Points.Count; i++)
        {
            Assert.Equal(resting[i], rope.Points[i].Position);
        }
    }

    [Fact]
    public void DraggingWakesTheRope()
    {
        var rope = MakeRope();
        Run(rope, 40.0);
        Assert.True(rope.IsSleeping);

        rope.BeginDrag(rope.Points[20].Position);

        Assert.False(rope.IsSleeping);
        rope.UpdateDrag(rope.Points[20].Position + new Vector2D(40, 0), new Vector2D(400, 0));
        rope.Step(Frame120);
        Assert.True(rope.LastStepCount > 0);
    }

    [Fact]
    public void ResizingWakesTheRope()
    {
        var rope = MakeRope();
        Run(rope, 40.0);
        Assert.True(rope.IsSleeping);

        rope.Resize(new Size(700, 300));

        Assert.False(rope.IsSleeping);
    }

    // MARK: - Reduce Motion

    [Fact]
    public void RestPoseIsStill()
    {
        var rope = MakeRope();
        rope.ResetToHanging();

        foreach (var point in rope.Points)
        {
            Assert.True(Math.Abs(point.Position.X - _anchor.X) < 1e-9);
            Assert.True(point.Displacement.Magnitude < 1e-9);
        }
        Assert.True(rope.Points[20].Position.Y > _anchor.Y + (rope.Configuration.TotalLength * 0.99));

        Run(rope, 2.0);
        Assert.True(rope.IsSleeping);
        Assert.True(Math.Abs(rope.Points[20].Position.X - _anchor.X) < 0.01);
    }

    // MARK: - Resizing

    [Fact]
    public void ResizingKeepsTheRopeAlive()
    {
        var rope = MakeRope();
        Run(rope, 1.0);

        rope.Resize(new Size(1040, 600));

        Assert.Equal(520.0, rope.Anchor.X);
        Assert.Equal(21, rope.Points.Count);
        Assert.Equal(20, rope.Configuration.SegmentCount);

        Run(rope, 3.0);
        Assert.True(rope.MeasuredMaximumStretch <= rope.Configuration.MaxStretchRatio + 1e-9);
    }
}

File: macos/Tests/RopeSimulationTests.swift
//
//  RopeSimulationTests.swift
//  HanglyTests
//

import CoreGraphics
import Foundation
import Testing

@testable import Hangly

/// Exercises the solver directly. The rope is AppKit-free, so every claim about it
/// can be checked numerically rather than by watching the screen.
@Suite("Rope simulation")
@MainActor
struct RopeSimulationTests {
    private let anchor = CGPoint(x: 260, y: 15)
    private let frame120: TimeInterval = 1.0 / 120.0

    private func makeRope() -> RopeSimulation {
        let rope = RopeSimulation(configuration: .default, anchor: anchor)
        rope.start()
        return rope
    }

    /// Advances `seconds` of simulated time in 120 Hz frames.
    private func run(_ rope: RopeSimulation, seconds: TimeInterval) {
        for _ in 0..<Int(seconds / frame120) {
            rope.step(deltaTime: frame120)
        }
    }

    // MARK: - Structure

    @Test("Twenty segments means twenty-one nodes")
    func segmentCountMatchesSpecification() {
        let rope = makeRope()

        #expect(rope.configuration.segmentCount == 20)
        #expect(rope.points.count == 21)
    }

    @Test("The first node is pinned and the charm is the heaviest")
    func massDistributionIsCorrect() {
        let rope = makeRope()

        #expect(rope.points[0].isPinned)
        #expect(rope.points[0].inverseMass == 0)
        // A heavier charm has a smaller inverse mass than a plain node.
        #expect(rope.points[20].inverseMass < rope.points[10].inverseMass)
        #expect(rope.points[10].inverseMass == 1)
    }

    @Test("The anchor never moves, however hard the rope is driven")
    func anchorStaysPinned() {
        let rope = makeRope()
        rope.beginDrag(at: rope.points[20].position)
        rope.updateDrag(to: CGPoint(x: 4000, y: -4000), velocity: CGPoint(x: 9000, y: -9000))
        run(rope, seconds: 2)

        #expect(rope.points[0].position == anchor)
    }

    // MARK: - Inextensibility

    @Test("The rope never stretches beyond its limit at rest")
    func doesNotStretchAtRest() {
        let rope = makeRope()
        run(rope, seconds: 5)

        #expect(rope.measuredMaximumStretch <= rope.configuration.maxStretchRatio + 1e-9)
    }

    @Test("The rope never stretches beyond its limit under a hard flick")
    func doesNotStretchUnderHardFlick() {
        let rope = makeRope()
        run(rope, seconds: 1)
        rope.beginDrag(at: rope.points[20].position)

        var worstStretch = 0.0
        for tick in 0..<600 {
            let angle = sin(Double(tick) * 0.1) * 0.3
            let target = anchor + CGPoint(x: sin(angle) * 200.0, y: cos(angle) * 200.0)
            rope.updateDrag(to: target, velocity: CGPoint(x: cos(angle) * 600.0, y: 0))
            rope.step(deltaTime: frame120)
            worstStretch = max(worstStretch, rope.measuredMaximumStretch)
        }

        #expect(worstStretch <= rope.configuration.maxStretchRatio + 1e-9)
    }

    @Test("Free swinging produces essentially no stretch")
    func doesNotStretchWhileSwinging() {
        let rope = makeRope()
        run(rope, seconds: 1)

        var worstStretch = 0.0
        for _ in 0..<2400 {
            rope.step(deltaTime: frame120)
            worstStretch = max(worstStretch, rope.measuredMaximumStretch)
        }

        #expect(worstStretch < 1.01)
    }

    @Test("Synthetic torture stays bounded and recovers")
    func recoversFromUnreachableInput() {
        let rope = makeRope()
        run(rope, seconds: 1)
        rope.beginDrag(at: rope.points[20].position)

        var worstStretch = 0.0
        for tick in 0..<600 {
            let angle = Double(tick) * 0.35
            let target = anchor + (CGPoint(x: cos(angle), y: sin(angle)) * 900)
            rope.updateDrag(to: target, velocity: CGPoint(x: 6000, y: 6000))
            rope.step(deltaTime: frame120)
            worstStretch = max(worstStretch, rope.measuredMaximumStretch)
        }
        #expect(worstStretch > 1.5)

        rope.endDrag()
        run(rope, seconds: 2)
        #expect(rope.measuredMaximumStretch <= rope.configuration.maxStretchRatio + 1e-9)
    }

    @Test("String stretches significantly on drag and oscillates back on release")
    func stringStretchesOnDragAndOscillatesOnRelease() {
        let rope = makeRope()
        run(rope, seconds: 2)
        rope.beginDrag(at: rope.points[20].position)

        let stretchedTarget = anchor + CGPoint(x: 0, y: 500)
        for _ in 0..<30 {
            rope.updateDrag(to: stretchedTarget, velocity: .zero)
            rope.step(deltaTime: frame120)
        }

        #expect(rope.measuredMaximumStretch > 2.0)

        rope.endDrag()
        rope.step(deltaTime: frame120)
        #expect(rope.points[20].displacement.y < 0)

        run(rope, seconds: 3)
        #expect(rope.measuredMaximumStretch <= rope.configuration.maxStretchRatio + 1e-9)
    }

    // MARK: - Gravity and damping

    @Test("Gravity swings the rope down below its anchor")
    func gravityPullsTheRopeDown() {
        let rope = makeRope()
        let startX = rope.points[20].position.x
        run(rope, seconds: 20)
        let charm = rope.points[20].position

        // It starts off-vertical by `initialAngle` and must end hanging under the anchor.
        #expect(abs(charm.x - anchor.x) < abs(startX - anchor.x) * 0.25)
        #expect(charm.y > anchor.y + (rope.configuration.totalLength * 0.9))
    }

    @Test("Damping brings the rope to rest")
    func dampingSettlesTheRope() {
        let rope = makeRope()
        run(rope, seconds: 3)
        let movingSpeed = totalSpeed(of: rope)

        run(rope, seconds: 25)
        let settledSpeed = totalSpeed(of: rope)

        #expect(settledSpeed < movingSpeed * 0.1)
    }

    private func totalSpeed(of rope: RopeSimulation) -> Double {
        rope.points.reduce(0) { $0 + $1.displacement.magnitude }
    }

    // MARK: - Frame-rate independence

    @Test("Two 120 Hz frames match one 60 Hz frame exactly")
    func fixedTimeStepMakesRefreshRateIrrelevant() {
        let fast = RopeSimulation(configuration: .default, anchor: anchor)
        let slow = RopeSimulation(configuration: .default, anchor: anchor)
        fast.start()
        slow.start()

        for _ in 0..<120 {
            fast.step(deltaTime: 1.0 / 120.0)
            fast.step(deltaTime: 1.0 / 120.0)
            slow.step(deltaTime: 1.0 / 60.0)
        }

        for index in fast.points.indices {
            #expect(abs(fast.points[index].position.x - slow.points[index].position.x) < 1e-9)
            #expect(abs(fast.points[index].position.y - slow.points[index].position.y) < 1e-9)
        }
    }

    @Test("A long 120 Hz run stays finite and bounded")
    func remainsStableOverALongRun() {
        let rope = makeRope()
        run(rope, seconds: 120)

        let reach = rope.configuration.totalLength * 3
        for point in rope.points {
            #expect(point.position.x.isFinite)
            #expect(point.position.y.isFinite)
            #expect(point.position.distance(to: anchor) < reach)
        }
    }

    @Test("A stalled frame cannot trigger a burst of catch-up steps")
    func clampsOversizedFrames() {
        let rope = makeRope()
        rope.step(deltaTime: 10)

        let ceiling = Int(rope.configuration.maxFrameDuration / rope.configuration.fixedTimeStep) + 1
        #expect(rope.lastStepCount <= ceiling)
    }
}

